src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -529,205 +538,177 @@ void checkConstraintLanguageVersionAndRuntimeSupportForConversion(SyntaxNode syn
-            BoundUnconvertedCollectionExpression node,
-            TypeSymbol targetType,
-            bool wasCompilerGenerated,
-            CollectionExpressionTypeKind collectionTypeKind,
-            TypeSymbol elementType,
-            BindingDiagnosticBag diagnostics)
-        {
-            var syntax = node.Syntax;
+            var syntax = (ExpressionSyntax)node.Syntax;
+            MethodSymbol? collectionBuilderMethod = null;
 
             switch (collectionTypeKind)
             {
                 case CollectionExpressionTypeKind.Span:
                     _ = GetWellKnownTypeMember(WellKnownMember.System_Span_T__ctor_Array, diagnostics, syntax: syntax);
                     break;
+
                 case CollectionExpressionTypeKind.ReadOnlySpan:
                     _ = GetWellKnownTypeMember(WellKnownMember.System_ReadOnlySpan_T__ctor_Array, diagnostics, syntax: syntax);
                     break;
+
+                case CollectionExpressionTypeKind.CollectionBuilder:
+                    {
+                        Debug.Assert(elementType is { });
+
+                        var namedType = (NamedTypeSymbol)targetType;
+                        bool result = namedType.HasCollectionBuilderAttribute(out TypeSymbol? builderType, out string? methodName);
+                        Debug.Assert(result);
+
+                        var targetTypeOriginalDefinition = targetType.OriginalDefinition;
+                        result = TryGetCollectionIterationType(syntax, targetTypeOriginalDefinition, out TypeWithAnnotations elementTypeOriginalDefinition);
+                        Debug.Assert(result);
+
+                        var useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+                        collectionBuilderMethod = GetCollectionBuilderMethod(namedType, elementTypeOriginalDefinition.Type, builderType, methodName, ref useSiteInfo);
+                        diagnostics.Add(syntax, useSiteInfo);
+                        if (collectionBuilderMethod is null)
+                        {
+                            diagnostics.Add(ErrorCode.ERR_CollectionBuilderAttributeMethodNotFound, syntax, methodName ?? "", elementTypeOriginalDefinition, targetTypeOriginalDefinition);
+                            return BindCollectionExpressionForErrorRecovery(node, targetType, diagnostics);
+                        }
+
+                        ReportUseSite(collectionBuilderMethod, diagnostics, syntax.Location);
+
+                        var parameterType = (NamedTypeSymbol)collectionBuilderMethod.Parameters[0].Type;
+                        Debug.Assert(parameterType.OriginalDefinition.Equals(Compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T), TypeCompareKind.AllIgnoreOptions));
+
+                        elementType = parameterType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0].Type;
+
+                        collectionBuilderMethod.CheckConstraints(
```

Comment:
Don't we need to construct the `collectionBuilderMethod` with the actual element type before checking constraints?

Reply:
The `collectionBuilderMethod` returned from `GetCollectionBuilderMethod()` has been constructed with the element type.

Reply:
maybe the `namedType` going in to `GetCollectionBuilderMethod` is constructed so this method is actually constructed with the right element type already?